prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -628,55 +660,59 @@ def build_tower(series):
         ),
     )
 
-    default_thermal_ratings = (
-        tower_designs.loc[default_designs]
-        .reset_index()
-        .set_index("voltage")
-        .apply(
-            lambda x: Line(tower=x["Tower"], voltage=x.name, length=1).thermal_rating,
-            axis=1,
-        )
-    )
-    # Now that we have Line objects for each Line, we can use the lower-level functions
-    branch["x"] = branch_plus_lines.apply(
-        lambda x: estimate_branch_impedance(x, bus_voltages), axis=1
-    )
-    branch["rateA"] = branch_plus_lines.apply(
-        lambda x: estimate_branch_rating(x, bus_voltages, default_thermal_ratings),
+    # Now that we have Lines for each Line, we can use the lower-level attributes
+    # Values for transmission lines ('type' == 'Line') can use their Line objects
+    line_voltages = branch_plus_line_designs["VOLTAGE"].unique()
+    z_bases = {v: calculate_z_base(v, const.s_base) for v in line_voltages}
+    branch["x"] = branch_plus_line_designs.apply(
+        lambda x: translate_to_per_unit(
+            x["line_object"].series_impedance.imag, "x", z_bases[x["VOLTAGE"]]
+        ),
         axis=1,
     )
+    branch["rateA"] = branch_plus_line_designs["line_object"].map(
+        lambda x: x.power_rating
+    )
 
 
-def estimate_branch_impedance(branch, bus_voltages):
-    """Estimate branch impedance using transformer voltages or line voltage and length.
+def estimate_transformers(bus_pair, lines, bus_voltages, transformer_designs):
+    """Determine the design of a transformer between two buses, and how many of them
+    should be added based on the connected lines.
 
-    :param pandas.Series branch: data for a single branch (line or transformer). All
-        branches require 'type' attributes, lines require 'VOLTAGE' and 'line_object',
-        transformers require 'from_bus_id' and 'to_bus_id'.
-    :param pandas.Series bus_voltages: mapping of buses to voltages.
-    :return: (*float*) -- impedance for that branch (per-unit).
+    :param iterable bus_pair: a pair of bus IDs, corresponding to indices in
+        ``bus_voltages`` and entries in the 'from_bus_id' and 'to_bus_id' columns of
+        ``lines``.
+    :param pandas.DataFrame lines: data frame of transmission lines containing columns:
+        'from_bus_id', 'to_bus_id', and 'rateA'.
+    :param pandas.Series bus_voltages: index is bus IDs, values are voltages (kV).
+    :param pandas.DataFrame transformer_designs: index is a tuple of voltage pairs in
+        (low_kV, high_kV) order, columns are 'x', 'r', and 'MVA'.
+    :return: (*pandas.Series*) -- indices are: 'x', 'r', 'MVA', and 'number'.
```

Comment:
values are 'number'?

Reply:
It's a Series with a single (not meaningfully ordered) index `['x', 'r', 'MVA', 'number']`. Essentially a dictionary.

Reply:
Got it. It's actually a series formed by a single row of a data frame.